File: mac/Sources/DMShot/Changelog.swift
import Foundation

struct ChangelogEntry: Equatable { let kind: String; let text: String }
struct ChangelogVersion: Equatable { let version: String; let date: String; let entries: [ChangelogEntry] }

/// Parses CHANGELOG.md (see Global Constraints for the format). Pure — no AppKit/Sparkle.
enum Changelog {
    private static let knownKinds: Set<String> = ["feat", "fix", "perf", "refactor", "docs", "chore"]

    static func parse(_ markdown: String) -> [ChangelogVersion] {
        var versions: [ChangelogVersion] = []
        var version: String?
        var date = ""
        var entries: [ChangelogEntry] = []

        func flush() {
            if let v = version { versions.append(ChangelogVersion(version: v, date: date, entries: entries)) }
        }

        for raw in markdown.split(separator: "\n", omittingEmptySubsequences: false) {
            let line = raw.trimmingCharacters(in: .whitespaces)
            if line.hasPrefix("## ") {
                flush()
                let header = line.dropFirst(3).trimmingCharacters(in: .whitespaces)
                // version and date are separated by an en-dash; dates use ASCII hyphens.
                let parts = header.components(separatedBy: "–").map { $0.trimmingCharacters(in: .whitespaces) }
                version = parts.first.flatMap { $0.isEmpty ? nil : $0 } ?? header
                date = parts.count > 1 ? parts[1] : ""
                entries = []
            } else if line.hasPrefix("- "), version != nil {
                let body = String(line.dropFirst(2))
                if let colon = body.firstIndex(of: ":") {
                    let kind = String(body[..<colon]).trimmingCharacters(in: .whitespaces).lowercased()
                    if knownKinds.contains(kind) {
                        let text = String(body[body.index(after: colon)...]).trimmingCharacters(in: .whitespaces)
                        entries.append(ChangelogEntry(kind: kind, text: text))
                        continue
                    }
                }
                entries.append(ChangelogEntry(kind: "other", text: body))
            }
        }
        flush()
        return versions
    }

    /// Release notes to show for an offered `version`: that version's entries if the bundled
    /// changelog has them, otherwise just the most recent version that has content. Empty
    /// placeholder sections (e.g. "[Unreleased]") are never shown. Keeps the Updates pane to
    /// the latest changes rather than the whole history — the installed build's changelog never
    /// contains the newer offered version, so an exact match usually fails and we'd otherwise
    /// dump everything.
    static func notes(_ all: [ChangelogVersion], for version: String) -> [ChangelogVersion] {
        let withContent = all.filter { !$0.entries.isEmpty }
        let matched = withContent.filter { $0.version == version }
        if !matched.isEmpty { return matched }
        return withContent.isEmpty ? [] : [withContent[0]]
    }

    /// Load + parse the bundled CHANGELOG.md (empty if missing — e.g. unbundled `swift run`).
    static func bundled(_ bundle: Bundle = .main) -> [ChangelogVersion] {
        guard let url = bundle.url(forResource: "CHANGELOG", withExtension: "md"),
              let text = try? String(contentsOf: url, encoding: .utf8) else { return [] }
        return parse(text)
    }
}

File: windows/DMShot/Update/Changelog.cs
using System.IO;
using System.Linq;
namespace DMShot.Update;

public readonly record struct ChangelogEntry(string Kind, string Text);
public sealed record ChangelogVersion(string Version, string Date, IReadOnlyList<ChangelogEntry> Entries);

/// <summary>
/// Parses CHANGELOG.md into versions + typed entries. Pure (no UI/Velopack), a direct
/// port of the macOS app's Changelog.swift so both platforms render "What's new" identically.
/// Format: "## &lt;version&gt; – &lt;date&gt;" headers (en-dash separator) with "- kind: text" bullets.
/// </summary>
public static class Changelog
{
    private static readonly HashSet<string> KnownKinds =
        new(StringComparer.OrdinalIgnoreCase) { "feat", "fix", "perf", "refactor", "docs", "chore" };

    public static IReadOnlyList<ChangelogVersion> Parse(string markdown)
    {
        var versions = new List<ChangelogVersion>();
        string? version = null;
        string date = "";
        var entries = new List<ChangelogEntry>();

        void Flush()
        {
            if (version is not null)
                versions.Add(new ChangelogVersion(version, date, entries));
        }

        foreach (var raw in markdown.Replace("\r\n", "\n").Split('\n'))
        {
            var line = raw.Trim();
            if (line.StartsWith("## "))
            {
                Flush();
                var header = line[3..].Trim();
                // version and date are separated by an en-dash; dates use ASCII hyphens.
                var parts = header.Split('–');
                var first = parts[0].Trim();
                version = first.Length == 0 ? header : first;
                date = parts.Length > 1 ? parts[1].Trim() : "";
                entries = new List<ChangelogEntry>();
            }
            else if (line.StartsWith("- ") && version is not null)
            {
                var body = line[2..];
                int colon = body.IndexOf(':');
                if (colon >= 0)
                {
                    var kind = body[..colon].Trim().ToLowerInvariant();
                    if (KnownKinds.Contains(kind))
                    {
                        var text = body[(colon + 1)..].Trim();
                        entries.Add(new ChangelogEntry(kind, text));
                        continue;
                    }
                }
                entries.Add(new ChangelogEntry("other", body));
            }
        }
        Flush();
        return versions;
    }

    /// <summary>
    /// Release notes to show for an offered <paramref name="version"/>: that version's entries
    /// if the bundled changelog has them, otherwise just the most recent version that has
    /// content. Empty placeholder sections (e.g. "[Unreleased]") are never shown. This keeps
    /// the Updates pane to the latest changes rather than the whole history — the installed
    /// build's changelog never contains the newer offered version, so an exact match usually
    /// fails and we'd otherwise dump everything.
    /// </summary>
    public static IReadOnlyList<ChangelogVersion> NotesFor(IReadOnlyList<ChangelogVersion> all, string version)
    {
        var withContent = all.Where(v => v.Entries.Count > 0).ToList();
        var matched = withContent.Where(v => v.Version == version).ToList();
        if (matched.Count > 0) return matched;
        return withContent.Count > 0 ? withContent.GetRange(0, 1) : new List<ChangelogVersion>();
    }

    /// <summary>Load + parse the CHANGELOG.md bundled next to the executable (empty if missing).</summary>
    public static IReadOnlyList<ChangelogVersion> Bundled()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "CHANGELOG.md");
        if (!File.Exists(path)) return new List<ChangelogVersion>();
        try { return Parse(File.ReadAllText(path)); }
        catch { return new List<ChangelogVersion>(); }
    }
}
